### DISCUSSION TITLE
Type breaks for curried create when using persist middleware since 4.4.5

### DISCUSSION THREAD

**Discussion by emanuel-lundman** [pmndrs/zustand#2163] [Bug report] [answered]
Noticed upgrading to 4.4.5 breaks our zustand context with type errors when using persist middleware.

`TimeLine` is our state in this case:
```
 Argument of type 'StateCreator<TimeLine, [], [["zustand/persist", TimeLine]]>' is not assignable to parameter of type 'StateCreator<TimeLine, [], [never, unknown][]>'.
  Type 'StateCreator<TimeLine, [], [["zustand/persist", TimeLine]]>' is not assignable to type '{ $$storeMutators?: [never, unknown][] | undefined; }'.
    Types of property '$$storeMutators' are incompatible.
      Type '[["zustand/persist", TimeLine]] | undefined' is not assignable to type '[never, unknown][] | undefined'.
```

errors out on persist:
![image](https://github.com/pmndrs/zustand/assets/2368897/f30bd407-d0b0-485c-bdbc-699b655ee29e)

**Comment by dai-shi** [comment]
That's unexpected.

> Types of property '$$storeMutators' are incompatible.

Dual package hazard??

Can you reproduce it in https://www.typscriptlang.org/play ?

**Comment by rffl** [comment]
Got the same issue.
Easiest way to reproduce:

Create a new file `fish.ts`
Paste the code from the docs
```

import { create } from 'zustand'
import { persist, createJSONStorage } from 'zustand/middleware'

const useFishStore = create(
  persist(
    (set, get) => ({
      fishes: 0,
      addAFish: () => set({ fishes: get().fishes + 1 }),
    }),
    {
      name: 'food-storage', // name of the item in the storage (must be unique)
      storage: createJSONStorage(() => sessionStorage), // (optional) by default, 'localStorage' is used
    }
  )
)
```

Got the error

```
No overload matches this call.
  Overload 1 of 3, '(initializer: StateCreator<never, [], [never, unknown][]>): UseBoundStore<StoreApi<never>>', gave the following error.
    Argument of type 'StateCreator<never, [], [["zustand/persist", unknown]]>' is not assignable to parameter of type 'StateCreator<never, [], [never, unknown][]>'.
      Type 'StateCreator<never, [], [["zustand/persist", unknown]]>' is not assignable to type '{ $$storeMutators?: [never, unknown][] | undefined; }'.
        Types of property '$$storeMutators' are incompatible.
          Type '[["zustand/persist", unknown]] | undefined' is not assignable to type '[never, unknown][] | undefined'.
            Type '[["zustand/persist", unknown]]' is not assignable to type '[never, unknown][]'.
              Type '["zustand/persist", unknown]' is not assignable to type '[never, unknown]'.
                Type at position 0 in source is not compatible with type at position 0 in target.
                  Type 'string' is not assignable to type 'never'.
  Overload 2 of 3, '(store: StoreApi<unknown>): UseBoundStore<StoreApi<unknown>>', gave the following error.
    Argument of type 'StateCreator<never, [], [["zustand/persist", unknown]]>' is not assignable to parameter of type 'StoreApi<unknown>'.
```

**Comment by dai-shi** [comment]
Can anyone identify which of the changes causes the issue?
- fix: importing CJS React in ESM by @dai-shi in https://github.com/pmndrs/zustand/pull/2154
- Apply publint recommendations by @dbritto-dev in https://github.com/pmndrs/zustand/pull/2157

**Reply by dbritto-dev** [reply]
@dai-shi I'll look into that

**Reply by dbritto-dev** [reply, marked as answer]
@dai-shi there's an error with `types: "./index.d.ts"` -> https://github.com/pmndrs/zustand/pull/2170

**Reply by dai-shi** [reply]
Thanks for the investigation! Though, I'm not sure why it caused such an issue. It feel like I misunderstand how "exports" are evaluated.

Can you or anyone can create a small reproduction with https://codesandbox.io/s/react-new ?

**Reply by dai-shi** [reply]
It doesn't seem that we can reproduce it in codesandbox, right?

**Reply by alexanderhorner** [reply]
Here is my [reproduction in codesandbox](https://codesandbox.io/p/sandbox/vibrant-bash-prcf7q?file=%2Fsrc%2Fstore.ts%3A13%2C1).

It only occurs when `"moduleResolution": "bundler",` is set in tsconfig.

**Reply by dai-shi** [reply]
"bundler", I see. Thanks!

**Comment by dai-shi** [comment]
Related discussions:
- #2164
- #2165
- #2169 
- #2171 
- #2172

**Comment by Athrun-Judah** [comment]
focus

**Comment by dai-shi** [comment]
Can anyone try this build in your app and give us a good sign?
https://ci.codesandbox.io/status/pmndrs/zustand/pr/2170/builds/434356
See "Local Install Instructions".

**Reply by alexanderhorner** [reply]
https://pkg.csb.dev/pmndrs/zustand/commit/bf044d7e/zustand does indeed fix the issue for me.

**Reply by TimonPeng** [reply]
this build with devtools and immer works fine for me
